core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java

```diff
@@ -144,23 +170,31 @@ public long acquireExecutionMemory(long required, MemoryConsumer consumer) {
       // spilling, avoid to have too many spilled files.
       if (got < required) {
         // Call spill() on other consumers to release memory
+        // Sort the consumers according their memory usage. So we avoid spilling the same consumer
+        // which is just spilled in last few times and re-spilling on it will produce many small
+        // spill files.
+        List<MemoryConsumer> sortedList = new ArrayList<>();
         for (MemoryConsumer c: consumers) {
           if (c != consumer && c.getUsed() > 0 && c.getMode() == mode) {
-            try {
-              long released = c.spill(required - got, consumer);
-              if (released > 0) {
-                logger.debug("Task {} released {} from {} for {}", taskAttemptId,
-                  Utils.bytesToString(released), c, consumer);
-                got += memoryManager.acquireExecutionMemory(required - got, taskAttemptId, mode);
-                if (got >= required) {
-                  break;
-                }
+            sortedList.add(c);
+          }
+        }
+        Collections.sort(sortedList, new ConsumerComparator());
+        for (MemoryConsumer c: sortedList) {
+          try {
+            long released = c.spill(required - got, consumer);
+            if (released > 0) {
+              logger.debug("Task {} released {} from {} for {}", taskAttemptId,
+                Utils.bytesToString(released), c, consumer);
+              got += memoryManager.acquireExecutionMemory(required - got, taskAttemptId, mode);
+              if (got >= required) {
+                break;
               }
-            } catch (IOException e) {
-              logger.error("error while calling spill() on " + c, e);
-              throw new OutOfMemoryError("error while calling spill() on " + c + " : "
-                + e.getMessage());
             }
+          } catch (IOException e) {
+            logger.error("error while calling spill() on " + c, e);
+            throw new OutOfMemoryError("error while calling spill() on " + c + " : "
+              + e.getMessage());
           }
```

Comment:
I don't think the best or better strategy is to spill many smaller consumers  for a bigger memory request.

Too frequently spill small consumers should cause GC pressure as reported in #16387. Actually this PR is inspired by that to avoid the same consumer is spilled again and again even it only uses small memory.

The newest update is to spill the ceiling entry (the least consumer uses greater than or equal to the required memory).

For the same example, consumers = 50 MB, 100MB, 200 MB, 250MB, 500 MB, 1GB, it behaves like:

Required: 1.4 GB
Evict: 1GB, 500MB

Required: 300 MB
Evict: 500MB

Required: 400MB
Evict: 500MB

Required: 60 MB
Evict 100 MB

Required: 200 MB
Evict 200 MB

It is nearly the same as the result you showed. As I pointed in previous comment, the logic you showed actually can't produce that result.